### ISSUE TITLE
subtype mismatch in black_box

### ISSUE THREAD

**Issue by franziskuskiefer** [AeneasVerif/eurydice#37]
The following code 

```rust
fn b(x: &[u8]) -> [u8; 32] {
    [0u8; 32]
}

pub fn bb(x: &[u8]) -> [u8; 32] {
    core::hint::black_box(b(x))
}
```

produces the following error when running it through eurydice

```
subtype mismatch:
  uint8_t[32size_t] -> uint8_t[32size_t] (a.k.a. uint8_t[32size_t] -> uint8_t[32size_t]) vs:
  uint8_t[32size_t] -> uint8_t[32size_t] -> () (a.k.a. uint8_t[32size_t] -> uint8_t[32size_t] -> ())
```

**Comment by msprotz** [comment]
This feels almost identical to the fix I just pushed for slice_index -- https://github.com/AeneasVerif/eurydice/pull/93

(note to self for when I investigate this further)

**Comment by msprotz** [comment]
So the issue is that black_box is an external function, and as such will be emitted as a macro. Macros only have one signature, and for `black_box`, the signature of the macro is `<T> T -> T`.

Remember that Rust arrays are values, and generate copies; however, they are compiled as `T[N]` in C. This means we rewrite types, so that a function type that takes and returns an array (like `black_box`) becomes a function type that takes a source and a destination array, and returns unit.

This means that there is a mismatch because the call-site wants a function of type `uint8_t[32size_t] -> uint8_t[32size_t] -> ()` (after monomorphization and type rewriting) but black_box is an external that keeps the type `<T> T -> T`, which instantiates to uint8_t[32size_t] -> uint8_t[32size_t] in this case.

There are several options here:
1. monomorphize external function prototypes -- for instance, Eurydice might emit references to `black_box_u8[32]`, `black_box_[u8]`, and potentially a lot more variants to account for each type that `black_box` is used at -- then, the user would have potentially infinitely many macros to "fill out" to provide implementations
2. monomorphize external function prototypes, but only for arrays -- this means that there would be two variants of black_box generated internally; one would be `black_box<T> (dummy: T) -> T` and the other would be `black_box_array<T> (dummy: T) (dst: T) -> ()`, with the invariant that the latter is used only at call-sites that (after monomorphization) instantiate black_box with an array type
3. change the compilation scheme of Rust arrays: instead of `T[N]`, we generate `typedef struct { T data[N]; } array_of_T_length_N`; every `e[x]` becomes `e.data[x]`, and so on... (@ssyram proposal)

Reviewing the options:
1. to me, this is not viable because there would just be too many variants to fill out, and this makes the code not future-proof (i.e., we'd be constantly adding variants to eurydice_glue.h)
2. this is potentially better, and does not result in an explosion of prototypes; furthermore, only external (opaque) functions that are *actually* used with array types would require implementing another macro; however, this adds more complexity to the "pass by ref" phase
3. this makes things very uniform, at the expense of i) having a LOT of extra type definitions in scope and ii) making things even more verbose by changing every `e[i]` to `e.data[i]`, and the same for assignments -- I initially contemplated this but went against it because I was trying to generate idiomatic C code and I was afraid of generating potentially hundreds of typedefs for all variants of known arrays and lengths

A common advantage of both 2. and 3. is that they would perhaps even remove hacks like `slice_to_array2`: right now, we have a few "known" external declarations that we know must always have their types rewritten because they're always used with array types... 

Separately, a workaround (just for black_box) would be to simply add a definition of black_box in Builtin.ml that reveals it is the identity, possibly with a custom attribute `KRML_NO_OPTIMIZE` so that instances of black_box after monomorphization are marked with a macro that we then define to indicate to the compiler that no optimization is desired.

Let's continue the conversation here -- it would be good to document the thinking so that we can come back to it later. Very curious to hear your thoughts @ssyram (and others)

**Comment by ssyram** [comment]
Thanks for the in depth discussion! I think simpler is better so I would still in support of 3) for a couple of additional reasons: 

1) the current implementation seems to use quite some difficult codes to handle the conversion, which is not that easily maintainable, while making stuff simple and direct makes stuff a lot easier;
2) this also makes things more predictable by the users (especially those want to check the correspondence for more complex real world codes). The change from a function returning array to accepting the return value and the extra copying seems to introduce additional indirectness;
3) The potential tons of new definitions for monomorphized types and functions seem not evitable from Rust to C anyway --- and arrays do not seem to be the bottleneck of this --- I think real world codes do not use stack arrays this much? Stack spaces, in my mind, are valuable that the programmers do never want to have the headaches of Stack-overflow.

Thanks again for the detailed feedback and actually we will be very happy to implement the new definitions.

**Comment by ssyram** [comment]
Thanks for this! Yes, I and @Lin23299 can work on this. A prototype should be ready around next week.

**Comment by Nadrieril** [comment]
Monomorphizing in charon would indeed fix this, but that may still take some time. If you have a fix for now that would be easier.

**Comment by msprotz** [comment]
> Add the genetic template Arr<T,C> {...} such that all concrete type like arr$u32$5 are instances of the template. Hopefully this can fix the type inconsistent.

This is the right solution. I thought this was what you were going to do originally.

**Comment by Lin23299** [comment]
Thanks for the feedback! I'm going to implement this solution.

**Comment by Lin23299** [comment]
Hello guys! I've updated the implementation using generic struct for arrays in `Builtin.ml` like this :
```
let arr : K.lident = [ "Eurydice" ], "arr"
let mk_arr (t : K.typ) (cg: K.cg) : K.typ = K.TCgApp (K.TApp (arr, [ t ]), cg)
```
This type is used in `AstOfLlbc` for the target type of arrays in Llbc and works smoothly. The arrays can be return by-value instead of by-ref as additional argument. I've also fixed the translation of `from_fn` and `map` in `Cleanup2.ml` according to this change. Now `array.rs` and `array2d.rs` can be tested.

However, the update of builtin functions becomes a problem. It is easy to change the  signatures of these functions in `Builtin.ml`(e.g. `array_to_subslice_to`) by replacing `T*` with `arr<T,N>*`. But their implementations in `eurydice_glue.h` are hard to change. Because`arr` as a generic type is different with `slice`, it cannot be directly defined in C. 
For the following function as an example:
```
static inline void core_num__u32__to_be_bytes(uint32_t src, uint8_t dst[4]) {
  // TODO: why not store32_be?
  uint32_t x = htobe32(src);
  memcpy(dst, &x, 4);
}
```
The declaration generated from `test/const_generics.rs` is:
```
static inline Eurydice_arr_e9 core_num__u32_to_be_bytes(uint32_t x0);
```

Clearly these two do not match and the compile fails. However, I do not know how to change the implementation becuase of the return type. Especially,
1. It seems very hard to define a generic struct type as`typedef struct arr'{ T data[N];} Eurydice_arr` in C similar to the the def of `Eurydice_slice`. It should contain values of a generic length instead of a reference, that's our start point I think. Maybe this is possible using some kind of macro defs?
2. The generic type `struct arr {data: [T;N]}` defined in `AstOfLlbc.ml` as we are doing now is monomorphized into random names like `Eurydice_arr_e9` instead of a specific name like `Eurydice_arr_u8_4`. Thus it's hard to use it as the return type in `eurydice_glue.h`.

I wonder how do you think about this problem. Anyway it seems like a big change is needed to
support this generic `Arr<T,N>` in builtin functions. @ssyram

**Comment by msprotz** [comment]
1. Yes, it would be nice, but one of my goals it for Eurydice to provide base implementations of common modules from the standard library. These implementations should live in the Eurydice codebase, so that they are shared by all projects rather than everyone reimplementing their own versions. In that case, this means that Eurydice needs to decide whether they exist as "static inline" or "extern", otherwise, you need to maintain two versions. Right now, there's a hardcoded flag here: https://github.com/AeneasVerif/eurydice/blob/main/bin/main.ml#L93 which is equivalent to a default configuration file. If you wanted, we could do two things to make stuff more modular. 1) Split the eurydice_glue.h header into multiple sub-headers, and 2) remove these hardcoded options in favor of a default.yaml file that lives in the Eurydice repository. That way, if someone has a strong opinion, they can only include the sub-headers that they care for, and override the yaml config file. What do you think?
2. Correct, but with pretty-printing the types, we ended up with 250-character long type names, like libcrux_ml_kem_ind_cca_generate_keypair_unpacked__libcrux_ml_kem_vector_neon_vector_type_SIMD128Vector_libcrux_ml_kem_hash_functions_neon_Simd128Hash_2size_t_768size_t_1632size_t_800size_t_768size_t_3size_t_192size_t (and that's not even the worst!).

**Comment by ssyram** [comment]
Yes, I think the first solution is good! Let's do that! As for the second point, Yeah, that's an issue. We will discuss more to figure out a better way... At least hashing seems to create potential of naming conflict and cannot provide sufficient info about this type.

**Comment by ssyram** [comment]
For the second point, we have come up with some new ideas:

1. How about keeping the full Rust name on the comments for these structs when defined, so we know which item it corresponds to? and,
2. How about having a common and public way of hashing so that we have a reliable way of hashing and we can provide our implementations reliably with this hashing --- not relying on OCaml.

**Comment by msprotz** [comment]
> For the second point, we have come up with some new ideas:
> 
>     1. How about keeping the full Rust name on the comments for these structs when defined, so we know which item it corresponds to? and,
> 
>     2. How about having a common and public way of hashing so that we have a reliable way of hashing and we can provide our implementations reliably with this hashing --- not relying on OCaml.

Yes to both that sounds reasonable! Happy to review proposals.
